tests/Rule/OptionalRuleTest.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Validator\Tests\Rule;
+
+use PHPUnit\Framework\TestCase;
+use Yiisoft\Validator\DataSetInterface;
+use Yiisoft\Validator\Rule\Number;
+use Yiisoft\Validator\Rule\OptionalRule;
+use Yiisoft\Validator\ValidationContext;
+
+final class OptionalRuleTest extends TestCase
+{
+    public function testBase(): void
+    {
+        $rule = new OptionalRule(
+            new Number()
+        );
+
+        $context = new ValidationContext(
+            $this->createDataSet(['code' => '42']),
+            'code',
+        );
+
+        $this->assertTrue($rule->validate('42', $context)->isValid());
+    }
+
+    public function testEmpty(): void
+    {
+        $rule = new OptionalRule(
+            new Number()
+        );
+
+        $context = new ValidationContext(
+            $this->createDataSet(['code' => '']),
+            'code',
+        );
+
+        $this->assertTrue($rule->validate('', $context)->isValid());
+    }
+
+    public function testEmptyCallback(): void
```

Comment:
Use data provider to test multiple cases or split several assert into several test cases.

Reply:
I'm move optional to another PR...